File: Python/Graphs/BellmanFord.py
"""
Bellman-Ford Algorithm Implementation

Time Complexity: O(V * E)
Space Complexity: O(V)
"""

import sys

def bellman_ford(edges, vertices_count, start_node):
    distances = [float('inf')] * vertices_count
    distances[start_node] = 0

    # Relax all edges V - 1 times
    for _ in range(vertices_count - 1):
        for u, v, w in edges:
            if distances[u] != float('inf') and distances[u] + w < distances[v]:
                distances[v] = distances[u] + w

    # Check for negative-weight cycles
    for u, v, w in edges:
        if distances[u] != float('inf') and distances[u] + w < distances[v]:
            return "Graph contains a negative weight cycle"

    return distances

# Example usage:
# if __name__ == "__main__":
#     # (source, destination, weight)
#     edges = [
#         (0, 1, -1), (0, 2, 4), (1, 2, 3), (1, 3, 2),
#         (1, 4, 2), (3, 2, 5), (3, 1, 1), (4, 3, -3)
#     ]
#     print(bellman_ford(edges, 5, 0)) # Output: [0, -1, 2, -2, 1]

File: TypeScript/Graphs/BellmanFord.ts
/**
 * Bellman-Ford Algorithm Implementation
 * 
 * Time Complexity: O(V * E)
 * Space Complexity: O(V)
 * 
 * Note: Can handle negative weights and detect negative cycles.
 */

export interface Edge {
    source: number;
    destination: number;
    weight: number;
}

export function bellmanFord(edges: Edge[], verticesCount: number, startNode: number): number[] | string {
    const distances = new Array(verticesCount).fill(Infinity);
    distances[startNode] = 0;

    // Relax all edges V - 1 times
    for (let i = 0; i < verticesCount - 1; i++) {
        for (const { source, destination, weight } of edges) {
            if (distances[source] !== Infinity && distances[source] + weight < distances[destination]) {
                distances[destination] = distances[source] + weight;
            }
        }
    }

    // Check for negative-weight cycles
    for (const { source, destination, weight } of edges) {
        if (distances[source] !== Infinity && distances[source] + weight < distances[destination]) {
            return "Graph contains a negative weight cycle";
        }
    }

    return distances;
}

// Example usage:
/*
const edges: Edge[] = [
    { source: 0, destination: 1, weight: -1 },
    { source: 0, destination: 2, weight: 4 },
    { source: 1, destination: 2, weight: 3 },
    { source: 1, destination: 3, weight: 2 },
    { source: 1, destination: 4, weight: 2 },
    { source: 3, destination: 2, weight: 5 },
    { source: 3, destination: 1, weight: 1 },
    { source: 4, destination: 3, weight: -3 }
];
console.log(bellmanFord(edges, 5, 0)); // Output: [0, -1, 2, -2, 1]
*/
